File: backend/app/models/movie.py
from sqlalchemy import Table, Column, Integer, String, Text, ForeignKey
from sqlalchemy.orm import relationship
from app.database.db import Base

# Association tables
movie_genres = Table(
    "movie_genres",
    Base.metadata,
    Column("movie_id", ForeignKey("movies.id"), primary_key=True),
    Column("genre_id", ForeignKey("genres.id"), primary_key=True),
)

movie_actors = Table(
    "movie_actors",
    Base.metadata,
    Column("movie_id", ForeignKey("movies.id"), primary_key=True),
    Column("actor_id", ForeignKey("actors.id"), primary_key=True),
)

class Movie(Base):
    __tablename__ = "movies"

    id = Column(Integer, primary_key=True)
    title = Column(String, nullable=False)
    release_year = Column(Integer, nullable=False)

    director_id = Column(Integer, ForeignKey("directors.id"))

    director = relationship("Director", back_populates="movies")
    genres = relationship("Genre", secondary=movie_genres, back_populates="movies")
    actors = relationship("Actor", secondary=movie_actors, back_populates="movies")
    reviews = relationship("MovieReview", back_populates="movie", cascade="all, delete")

class Actor(Base):
    __tablename__ = "actors"

    id = Column(Integer, primary_key=True)
    name = Column(String, nullable=False)

    movies = relationship("Movie", secondary=movie_actors, back_populates="actors")

class Genre(Base):
    __tablename__ = "genres"

    id = Column(Integer, primary_key=True)
    name = Column(String, unique=True, nullable=False)
    description = Column(String(255), nullable=True)    
    slug = Column(String(50), unique=True, index=True)
    movies = relationship("Movie", secondary=movie_genres, back_populates="genres")

class Director(Base):
    __tablename__ = "directors"

    id = Column(Integer, primary_key=True)
    name = Column(String, nullable=False)

    movies = relationship("Movie", back_populates="director")

class MovieReview(Base):
    __tablename__ = "movie_reviews"

    id = Column(Integer, primary_key=True)
    movie_id = Column(Integer, ForeignKey("movies.id"), nullable=False)
    reviewer_name = Column(String, nullable=False)
    rating = Column(Integer, nullable=False)  # 1–5
    comment = Column(Text)

    movie = relationship("Movie", back_populates="reviews")

File: frontend/src/models/movie.ts
export interface Entity {
  id: number;
  name: string;
}

export interface MovieSummary {
  id: number;
  title: string;
}

export interface ActorProfileData extends Entity {
  movies: MovieSummary[];
}

export interface DirectorProfileData extends Entity {
  movies: MovieSummary[];
}

export interface Movie {
  id: number;
  title: string;
  release_year: number;
  director: Entity;
  genres: Entity[];
  actors: Entity[];
  reviews: Review[];
}

export interface MovieFilters {
  genre_id: string;
  actor_id: string;
  director_id: string;
}

export interface FilterBarProps {
  onFilter: (filters: MovieFilters) => void;
  onReady?: () => void;
}

export interface MovieCardProps {
  movie: Movie;
}

export interface Review {
  id: number;
  rating: number;
  comment: string;
} 